### DISCUSSION TITLE
pydantic in docs

### DISCUSSION THREAD

**Discussion by hyperknot** [litestar-org/litestar#2431] [General]
When reading the docs, I frequently encounter references to pydantic. If I understand it correctly, it's been replaced with msgspec (thank you for this, BTW!)

What's not clear is that the references in the docs can be understood with replace-all to "msgspec" or some are pydantic specific, and only work if pydantic is installed as an optional dependency?

For example https://docs.litestar.dev/latest/usage/routing/parameters.html has 11 occurances of "pydantic", but other docs pages also have about 3-4 on average.

**Comment by peterschutt** [comment]
We should probably just use dataclasses wherever we are arbitrarily using models in examples. As you know, pydantic was once a core dependency and these examples made more sense under those conditions.

**Reply by hyperknot** [reply]
But what is actually happening behind the scenes? date, etc. is converted using msgspec now?

**Reply by peterschutt** [reply]
Yes, pydantic is no longer used for parsing or validation of handler signature params/path params, etc.

**Reply by hyperknot** [reply]
But what is the recommendation, should I use dataclasses and forget about msgspec in user code?

As far as I can see, msgspec is kind of hidden behind the scenes, not even the performance-test code is using it. Which is strange, wouldn't it be faster if performance-test used it directly?

**Reply by hyperknot** [reply]
Here is a benchmark about msgspec.Struct:
https://jcristharif.com/msgspec/benchmarks.html#benchmark-structs

- Structs are optimized to be cheap to create, and that shows for the creation benchmark. They’re roughly 5x faster than standard classes/attrs/dataclasses, and 50x faster than pydantic.

- For equality comparison, msgspec Structs are roughly 6x to 500x faster than the alternatives.

- For order comparison, msgspec Structs are roughly 10x to 70x faster than the alternatives.

**Reply by peterschutt** [reply]
We don't want to proscribe what anyone should be using for their data modelling. We can natively support modelling with Structs and pydantic models, because msgspec natively supports them.

> As far as I can see, msgspec is kind of hidden behind the scenes,

Yes, we use it for parsing and validation of the request body, and parameters/headers etc.

**Reply by hyperknot** [reply]
I tried out a modified api-performance-test with Struct instead of dataclass and there wasn't meaningful difference in rps. So I guess probably dataclass is the nicest DX as it only uses stdlib.
